src/backend/xacheck.go

```diff
@@ -0,0 +1,330 @@
+	if xc == nil {
+		return nil
+	}
+
+	xaCommitErr := &XaCommitErr{
+		Time:     time.Now().Format("20060102150405"),
+		Xaid:     txn.xid,
+		State:    state,
+	}
+
+	xc.mu.Lock()
+	defer xc.mu.Unlock()
+	// add the xaCommitErrLog to xacheck
+	if err := xc.addXaCommitErrLog(xaCommitErr); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// TODO if the Radon crash at the moment.
+	// flush the mem to the file
+	if err := xc.flushXaCommitErrLog(); err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+// commitRetryBackends
+// 1st. xa recover to check the xid is valid to some backends which need the 2nd command
+// 2nd. xa commit/rollback 'xaid' to the Backends which need commit
+func (xc *XaCheck) commitRetryBackends(query string, scatter *Scatter, xid string) (bool, error) {
+	backends := scatter.Backends()
+	log := xc.log
+
+	// if the backend is empty, output warning log.
+	if len(backends) == 0 {
+		log.Error("xacheck.commitRetryBackends.backend.empty.")
+		return false, errors.New("xacheck.backend.empty")
+	}
+
+	// the 1st stage: xa recover
+	txn, err := scatter.CreateTransaction()
+	if err != nil {
+		log.Error("xacheck.commitRetryBackends.create.transaction.error:[%v]", err)
+		return false, err
+	}
+	defer txn.Finish()
+
+	xaRecoverQuery := "xa recover"
+	var needCommitBackends []string
+	for _, backend := range backends {
+		result, err := txn.ExecuteOnThisBackend(backend, xaRecoverQuery)
+		if err != nil {
+			continue
+		}
+
+		if result != nil && result.RowsAffected > 0 && len(result.Fields) == 4 {
+			for _, row := range(result.Rows) {
+				valStr := string(row[3].Raw())
+				if strings.EqualFold(valStr, xid) {
+					needCommitBackends = append(needCommitBackends, backend)
+				}
+			}
+		}
+	}
+
+	// the 2nd stage: xa commit/rollback '$xid'
+	ExecuteOKCount := 0
+	for _, backend := range needCommitBackends {
+		_, err = txn.ExecuteOnThisBackend(backend, query)
+		if err == nil {
+			log.Info("xacheck.commitRetryBackends.query[%v].success.backend[%v]", query, backend)
+			ExecuteOKCount++
+		} else {
+			log.Warning("xacheck.commitRetryBackends.query[%v].backend[%v].error[%T]:%+v", query, backend, err, err)
+		}
+	}
+
+	// if there are items need to retry, and ExecuteOKCount is equal to the needCommitBackends, return true
+	if ExecuteOKCount > 0 && ExecuteOKCount == len(needCommitBackends) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// xaCommitsRetryMain in which the retrys maybe reduce when the valid backends need commit and succeed
+func (xc *XaCheck) xaCommitsRetryMain() error {
+	log := xc.log
+	retrys := xc.retrys
+	if (len(retrys) > 0) {
+		log.Info("xacheck.commit.retry %v.", retrys)
+	}
+
+	for _, retry := range retrys {
+		query := fmt.Sprintf("xa %s '%s' ", retry.State, retry.Xaid)
+		committed, err := xc.commitRetryBackends(query, xc.scatter, retry.Xaid);
+		if err != nil {
+			log.Warning("xacheck.commits.retry failed.")
+			return err
+		}
+
+		if committed {
```

Comment:
As the above talked, if we have 5 backends, and backend3, backend4 'xa commit' error.
If the backend3 query 'xa recover' failed, the neeedCommit is 1 not 2, so the backend3 will never be retry the 'xa commit'.

Reply:
you are so serious and kind，tell me why and how. I will fix it and add some comments